ntbs-service/Services/AuthorizationService.cs

```diff
@@ -184,26 +187,33 @@ public async Task<IQueryable<Notification>> FilterNotificationsByUserAsync(Claim
 
         public async Task<IList<Alert>> FilterAlertsForUserAsync(ClaimsPrincipal user, IList<Alert> alerts)
         {
-            var userTbServiceCodes = (await _userService.GetTbServicesAsync(user)).Select(s => s.Code);
+            var userTbServiceCodes = (await _userService.GetTbServicesAsync(user)).Select(s => s.Code).ToList();
             var userType = _userService.GetUserType(user);
-            for (int i = alerts.Count - 1; i >= 0; i--)
+            alerts = alerts.Where(a => AlertIsForNotificationInOrTransferToGivenServices(a, userTbServiceCodes)).ToList();
+            if (userType == UserType.PheUser)
             {
-                // For transfer alerts PHE users belonging to a PHEC cannot see and action transfer alerts as they are
-                // aimed to be actioned on a TB service level
-                if (userType == UserType.PheUser && (alerts[i].AlertType == AlertType.TransferRequest ||
-                                                     alerts[i].AlertType == AlertType.TransferRejected))
-                {
-                    alerts.RemoveAt(i);
-                }
-                else if (alerts[i].TbServiceCode != null && !userTbServiceCodes.Contains(alerts[i].TbServiceCode))
-                    alerts.RemoveAt(i);
+                alerts = alerts.Where(a => a.AlertType != AlertType.TransferRequest && 
+                                           a.AlertType != AlertType.TransferRejected).ToList();
             }
-            return alerts;
+            return alerts.ToList();
         }
 
         private async Task<UserPermissionsFilter> GetUserPermissionsFilterAsync(ClaimsPrincipal user)
         {
             return await _userService.GetUserPermissionsFilterAsync(user);
         }
+
+        private bool AlertIsForNotificationInOrTransferToGivenServices(Alert alert, ICollection<string> givenTbServiceCodes)
+        {
+            var notificationIdsInServices = _notificationRepository.GetBaseNotificationsIQueryable()
+                .Where(n => givenTbServiceCodes.Contains(n.HospitalDetails.TBServiceCode))
+                .Select(n => n.NotificationId);

```

Comment:
It's weird that this logic is in the `AlertIsForNotificationInOrTransferToGivenServices` which will be called once for every alert in the list. If Entity Framework doesn't optimize this could end up being called many times.

Now, I _think_ that EF will be clever enough to cache the result of the query and re-use it, but I wouldn't be certain. Either way it would be clearer if we calculate this once and pass it in here.

Reply:
I can move this out, I think the only reason it's in here is because this function is called from `IsUserAuthorizedToManageAlert` as well as `FilterAlertsForUserAsync` so both will have to pass it in. But it makes sense not to call it for every alert